python/taichi/_kernels.py

```diff
@@ -241,9 +243,15 @@ def fill_matrix(mat: template(), vals: template()):
         for p in static(range(mat.n)):
             for q in static(range(mat.m)):
                 if static(mat[I].ndim == 2):
-                    mat[I][p, q] = vals[p][q]
+                    if static(isinstance(vals[p], Iterable)):
```

Comment:
Currently, there are `Matrix` and `_IntermediateMatrix`. The former one can be clearly told apart from `Vector`s; however, unfortunately, the latter one cannot be because there are code paths that constructs `_IntermediateMatrix` from `Vector`s. But after the refactor this could be purely distinguished by # of dimensions. Good point, thanks!